File: src/frontend/js/command/panels/annotations.js
// Created by Matthew Valancy
// Copyright 2026 Valpatel Software LLC
// Licensed under AGPL-3.0 — see LICENSE for details.
/**
 * Map Annotations Panel — place text labels, arrows, circles, and freehand
 * drawings on the tactical map. Annotations persist via REST API.
 *
 * Used for briefings, operational planning, and marking areas of interest.
 */

import { EventBus } from '../events.js';
import { TritiumStore } from '../store.js';

let _panel = null;
let _annotations = [];
let _activeLayer = 'default';
let _drawMode = null;  // null | 'text' | 'arrow' | 'circle' | 'freehand' | 'rectangle'
let _currentColor = '#00f0ff';
let _freehandPoints = [];

const ANNOTATION_COLORS = [
    '#00f0ff', '#ff2a6d', '#05ffa1', '#fcee0a',
    '#ff9800', '#ffffff', '#ff00ff', '#00a0ff',
];

// ---------------------------------------------------------------------------
// API
// ---------------------------------------------------------------------------

async function _fetchAnnotations() {
    try {
        const res = await fetch(`/api/annotations?layer=${encodeURIComponent(_activeLayer)}`);
        if (res.ok) {
            const data = await res.json();
            _annotations = data.annotations || [];
            EventBus.emit('annotations:updated', _annotations);
        }
    } catch (e) {
        console.warn('[ANNOTATIONS] fetch failed:', e);
    }
}

async function _createAnnotation(ann) {
    try {
        const res = await fetch('/api/annotations', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(ann),
        });
        if (res.ok) {
            const created = await res.json();
            _annotations.push(created);
            EventBus.emit('annotations:updated', _annotations);
            return created;
        }
    } catch (e) {
        console.warn('[ANNOTATIONS] create failed:', e);
    }
    return null;
}

async function _deleteAnnotation(id) {
    try {
        await fetch(`/api/annotations/${id}`, { method: 'DELETE' });
        _annotations = _annotations.filter(a => a.id !== id);
        EventBus.emit('annotations:updated', _annotations);
    } catch (e) {
        console.warn('[ANNOTATIONS] delete failed:', e);
    }
}

async function _clearLayer() {
    try {
        await fetch(`/api/annotations?layer=${encodeURIComponent(_activeLayer)}`, { method: 'DELETE' });
        _annotations = [];
        EventBus.emit('annotations:updated', _annotations);
    } catch (e) {
        console.warn('[ANNOTATIONS] clear failed:', e);
    }
}

// ---------------------------------------------------------------------------
// Drawing mode
// ---------------------------------------------------------------------------

function setDrawMode(mode) {
    _drawMode = mode;
    TritiumStore.set('annotations.drawMode', mode);
    EventBus.emit('annotations:mode', mode);
    _renderTools();
}

function getDrawMode() {
    return _drawMode;
}

function handleMapClick(lngLat) {
    if (!_drawMode) return false;

    if (_drawMode === 'text') {
        const text = prompt('Annotation text:');
        if (text) {
            _createAnnotation({
                type: 'text',
                lat: lngLat.lat,
                lng: lngLat.lng,
                text: text,
                color: _currentColor,
                layer: _activeLayer,
            });
        }
        setDrawMode(null);
        return true;
    }

    if (_drawMode === 'circle') {
        const radiusStr = prompt('Circle radius (meters):', '50');
        const radius = parseFloat(radiusStr);
        if (radius > 0) {
            _createAnnotation({
                type: 'circle',
                lat: lngLat.lat,
                lng: lngLat.lng,
                radius_m: radius,
                color: _currentColor,
                fill: true,
                layer: _activeLayer,
            });
        }
        setDrawMode(null);
        return true;
    }

    if (_drawMode === 'arrow') {
        if (!_panel._arrowStart) {
            _panel._arrowStart = { lat: lngLat.lat, lng: lngLat.lng };
            _renderTools();
            return true;
        } else {
            _createAnnotation({
                type: 'arrow',
                lat: _panel._arrowStart.lat,
                lng: _panel._arrowStart.lng,
                end_lat: lngLat.lat,
                end_lng: lngLat.lng,
                color: _currentColor,
                layer: _activeLayer,
            });
            _panel._arrowStart = null;
            setDrawMode(null);
            return true;
        }
    }

    if (_drawMode === 'freehand') {
        _freehandPoints.push([lngLat.lat, lngLat.lng]);
        return true;
    }

    return false;
}

function finishFreehand() {
    if (_freehandPoints.length >= 2) {
        _createAnnotation({
            type: 'freehand',
            lat: _freehandPoints[0][0],
            lng: _freehandPoints[0][1],
            points: _freehandPoints,
            color: _currentColor,
            layer: _activeLayer,
        });
    }
    _freehandPoints = [];
    setDrawMode(null);
}

// ---------------------------------------------------------------------------
// Panel rendering
// ---------------------------------------------------------------------------

function _renderTools() {
    if (!_panel) return;
    const toolbar = _panel.querySelector('.annotation-toolbar');
    if (!toolbar) return;

    const modes = [
        { id: 'text', icon: 'T', label: 'Text' },
        { id: 'arrow', icon: '\u2192', label: 'Arrow' },
        { id: 'circle', icon: '\u25CB', label: 'Circle' },
        { id: 'freehand', icon: '\u270E', label: 'Draw' },
    ];

    toolbar.innerHTML = modes.map(m => `
        <button class="ann-tool ${_drawMode === m.id ? 'active' : ''}"
                data-mode="${m.id}" title="${m.label}">
            <span class="ann-tool-icon">${m.icon}</span>
        </button>
    `).join('') + `
        <span class="ann-color-swatch" style="background:${_currentColor}" title="Color"></span>
        <button class="ann-tool" data-action="clear" title="Clear layer">
            <span class="ann-tool-icon">\u2715</span>
        </button>
    `;

    // Status line
    const status = _panel.querySelector('.annotation-status');
    if (status) {
        if (_drawMode === 'arrow' && _panel._arrowStart) {
            status.textContent = 'Click end point for arrow';
        } else if (_drawMode === 'freehand' && _freehandPoints.length > 0) {
            status.textContent = `Drawing... ${_freehandPoints.length} points (Esc to finish)`;
        } else if (_drawMode) {
            status.textContent = `Click map to place ${_drawMode}`;
        } else {
            status.textContent = `${_annotations.length} annotations on layer: ${_activeLayer}`;
        }
    }

    // Bind events
    toolbar.querySelectorAll('.ann-tool[data-mode]').forEach(btn => {
        btn.onclick = () => {
            const mode = btn.dataset.mode;
            setDrawMode(_drawMode === mode ? null : mode);
        };
    });
    toolbar.querySelectorAll('.ann-tool[data-action="clear"]').forEach(btn => {
        btn.onclick = () => {
            if (confirm('Clear all annotations on this layer?')) _clearLayer();
        };
    });
    const swatch = toolbar.querySelector('.ann-color-swatch');
    if (swatch) {
        swatch.onclick = () => {
            const idx = ANNOTATION_COLORS.indexOf(_currentColor);
            _currentColor = ANNOTATION_COLORS[(idx + 1) % ANNOTATION_COLORS.length];
            swatch.style.background = _currentColor;
        };
    }
}

function _renderList() {
    if (!_panel) return;
    const list = _panel.querySelector('.annotation-list');
    if (!list) return;

    if (_annotations.length === 0) {
        list.innerHTML = '<div class="ann-empty">No annotations on this layer</div>';
        return;
    }

    list.innerHTML = _annotations.map(a => `
        <div class="ann-item" data-id="${a.id}">
            <span class="ann-item-type" style="color:${a.color}">${a.type}</span>
            <span class="ann-item-text">${a.text || a.label || a.type}</span>
            <button class="ann-item-delete" data-id="${a.id}" title="Delete">\u2715</button>
        </div>
    `).join('');

    list.querySelectorAll('.ann-item-delete').forEach(btn => {
        btn.onclick = (e) => {
            e.stopPropagation();
            _deleteAnnotation(btn.dataset.id);
        };
    });
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

export function initAnnotationsPanel(container) {
    _panel = document.createElement('div');
    _panel.className = 'panel-content annotations-panel';
    _panel.innerHTML = `
        <div class="panel-header">MAP ANNOTATIONS</div>
        <div class="annotation-toolbar"></div>
        <div class="annotation-status"></div>
        <div class="annotation-list"></div>
    `;
    _panel._arrowStart = null;
    container.appendChild(_panel);

    _fetchAnnotations();
    _renderTools();

    EventBus.on('annotations:updated', () => {
        _renderList();
        _renderTools();
    });

    // Periodic refresh
    setInterval(_fetchAnnotations, 10000);

    return _panel;
}

export function destroyAnnotationsPanel() {
    if (_panel && _panel.parentNode) {
        _panel.parentNode.removeChild(_panel);
    }
    _panel = null;
}

export function getAnnotations() {
    return _annotations;
}

export { setDrawMode, getDrawMode, handleMapClick, finishFreehand };

// PanelDef for PanelManager registration
export const AnnotationsPanelDef = {
    id: 'annotations',
    label: 'ANNOTATIONS',
    icon: '\u270F',
    group: 'map',
    init(container) { return initAnnotationsPanel(container); },
    destroy() { destroyAnnotationsPanel(); },
};

File: src/app/routers/annotations.py
# Created by Matthew Valancy
# Copyright 2026 Valpatel Software LLC
# Licensed under AGPL-3.0 — see LICENSE for details.
"""Map annotation system — persist text labels, arrows, circles, and freehand
drawings on the tactical map for briefings and operational planning.

Annotations are stored in-memory (reset on server restart). Each annotation
has a type, position, style, and optional metadata.
"""

from __future__ import annotations

import time
import uuid
from typing import Any, Optional

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

router = APIRouter(prefix="/api/annotations", tags=["annotations"])


# ---------------------------------------------------------------------------
# Models
# ---------------------------------------------------------------------------

class AnnotationCreate(BaseModel):
    """Create a new map annotation."""
    type: str = Field(..., description="text | arrow | circle | freehand | rectangle | polygon")
    lat: float = Field(..., description="Latitude of annotation anchor")
    lng: float = Field(..., description="Longitude of annotation anchor")
    # Text content (for text annotations)
    text: str = ""
    # Geometry (type-specific)
    end_lat: Optional[float] = None  # arrow end point
    end_lng: Optional[float] = None
    radius_m: Optional[float] = None  # circle radius in meters
    points: Optional[list[list[float]]] = None  # freehand/polygon: [[lat,lng], ...]
    width: Optional[float] = None  # rectangle width in meters
    height: Optional[float] = None  # rectangle height in meters
    # Style
    color: str = "#00f0ff"
    stroke_width: float = 2.0
    font_size: float = 14.0
    opacity: float = 0.8
    fill: bool = False
    fill_opacity: float = 0.2
    # Metadata
    label: str = ""
    layer: str = "default"
    locked: bool = False


class AnnotationUpdate(BaseModel):
    """Partial update for an annotation."""
    lat: Optional[float] = None
    lng: Optional[float] = None
    text: Optional[str] = None
    end_lat: Optional[float] = None
    end_lng: Optional[float] = None
    radius_m: Optional[float] = None
    points: Optional[list[list[float]]] = None
    width: Optional[float] = None
    height: Optional[float] = None
    color: Optional[str] = None
    stroke_width: Optional[float] = None
    font_size: Optional[float] = None
    opacity: Optional[float] = None
    fill: Optional[bool] = None
    fill_opacity: Optional[float] = None
    label: Optional[str] = None
    layer: Optional[str] = None
    locked: Optional[bool] = None


# ---------------------------------------------------------------------------
# In-memory store
# ---------------------------------------------------------------------------

_annotations: dict[str, dict[str, Any]] = {}


# ---------------------------------------------------------------------------
# Endpoints
# ---------------------------------------------------------------------------

@router.get("")
async def list_annotations(layer: Optional[str] = None):
    """List all map annotations, optionally filtered by layer."""
    items = list(_annotations.values())
    if layer:
        items = [a for a in items if a.get("layer") == layer]
    return {
        "annotations": sorted(items, key=lambda a: a.get("created_at", 0)),
        "count": len(items),
    }


@router.post("")
async def create_annotation(body: AnnotationCreate):
    """Create a new map annotation."""
    now = time.time()
    ann_id = f"ann_{uuid.uuid4().hex[:8]}"
    annotation = {
        "id": ann_id,
        "type": body.type,
        "lat": body.lat,
        "lng": body.lng,
        "text": body.text,
        "end_lat": body.end_lat,
        "end_lng": body.end_lng,
        "radius_m": body.radius_m,
        "points": body.points,
        "width": body.width,
        "height": body.height,
        "color": body.color,
        "stroke_width": body.stroke_width,
        "font_size": body.font_size,
        "opacity": body.opacity,
        "fill": body.fill,
        "fill_opacity": body.fill_opacity,
        "label": body.label,
        "layer": body.layer,
        "locked": body.locked,
        "created_at": now,
        "updated_at": now,
    }
    _annotations[ann_id] = annotation
    return annotation


@router.get("/{annotation_id}")
async def get_annotation(annotation_id: str):
    """Get a single annotation by ID."""
    ann = _annotations.get(annotation_id)
    if ann is None:
        raise HTTPException(status_code=404, detail="Annotation not found")
    return ann


@router.put("/{annotation_id}")
async def update_annotation(annotation_id: str, body: AnnotationUpdate):
    """Update an existing annotation."""
    ann = _annotations.get(annotation_id)
    if ann is None:
        raise HTTPException(status_code=404, detail="Annotation not found")
    if ann.get("locked"):
        raise HTTPException(status_code=403, detail="Annotation is locked")

    updates = body.model_dump(exclude_none=True)
    if updates:
        ann.update(updates)
        ann["updated_at"] = time.time()

    return ann


@router.delete("/{annotation_id}")
async def delete_annotation(annotation_id: str):
    """Delete an annotation."""
    if annotation_id not in _annotations:
        raise HTTPException(status_code=404, detail="Annotation not found")
    del _annotations[annotation_id]
    return {"ok": True, "deleted": annotation_id}


@router.delete("")
async def clear_annotations(layer: Optional[str] = None):
    """Clear all annotations, optionally filtered by layer."""
    if layer:
        to_delete = [k for k, v in _annotations.items() if v.get("layer") == layer]
    else:
        to_delete = list(_annotations.keys())
    for k in to_delete:
        del _annotations[k]
    return {"ok": True, "deleted_count": len(to_delete)}


@router.get("/layers/list")
async def list_layers():
    """List all annotation layers."""
    layers = set()
    for ann in _annotations.values():
        layers.add(ann.get("layer", "default"))
    return {"layers": sorted(layers)}
